README
NAME
    DBIx::Class::Loader - Dynamic definition of DBIx::Class sub classes.

SYNOPSIS
      use DBIx::Class::Loader;

      my $loader = DBIx::Class::Loader->new(
        dsn                     => "dbi:mysql:dbname",
        user                    => "root",
        password                => "",
        namespace               => "Data",
        additional_classes      => qw/DBIx::Class::Foo/,
        additional_base_classes => qw/My::Stuff/,
        left_base_classes       => qw/DBIx::Class::Bar/,
        constraint              => '^foo.*',
        relationships           => 1,
        options                 => { AutoCommit => 1 }, 
        inflect                 => { child => 'children' }
      );
      my $class = $loader->find_class('film'); # $class => Data::Film
      my $obj = $class->retrieve(1);

    use with mod_perl

    in your startup.pl

      # load all tables
      use DBIx::Class::Loader;
      my $loader = DBIx::Class::Loader->new(
        dsn       => "dbi:mysql:dbname",
        user      => "root",
        password  => "",
        namespace => "Data",
      );

    in your web application.

      use strict;

      # you can use Data::Film directly
      my $film = Data::Film->retrieve($id);

DESCRIPTION
    DBIx::Class::Loader automate the definition of DBIx::Class sub-classes.
    scan table schemas and setup columns, primary key.

    class names are defined by table names and namespace option.

     +-----------+-----------+-----------+
     |   table   | namespace | class     |
     +-----------+-----------+-----------+
     |   foo     | Data      | Data::Foo |
     |   foo_bar |           | FooBar    |
     +-----------+-----------+-----------+

    DBIx::Class::Loader supports MySQL, Postgres and SQLite.

    See DBIx::Class::Loader::Generic for more, and
    DBIx::Class::Loader::Writing for notes on writing your own db-specific
    subclass for an unsupported db.

    Class::DBI::Loader and Class::DBI are now obsolete, use DBIx::Class and
    this module instead. ;)

AUTHOR
    Sebastian Riedel, "sri@oook.de"

    Based upon the work of IKEBE Tomohiro

THANK YOU
    Adam Anderson, Andy Grundman, Autrijus Tang, Dan Kubb, David Naughton,
    Randal Schwartz, Simon Flack and all the others who've helped.

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    DBIx::Class



Changes
Revision history for Perl extension DBIx::Class::Loader

0.20  Tue Feb 14 18:26:31 2006
        - Added correct bug contact info
        - Vendor modules now use ->storage

0.19  Sun Feb 12 22:24:32 2006
        - No longer tries to parse out the table name from the dsn for mysql,
          was unneccesary vestigial code from previous method.

0.18  Sat Feb 04 00:18:58 2006
        - Docs updates, including module deprecation notice

0.17  Fri Feb 03 01:45:27 2006
        - Set postgresql default schema back to 'public' instead of ''

0.16  Tue Jan 31 03:33:48 2006
        - Fixed tests on non-innodb mysql installations
        - do ->storage->dbh->disconnect when done loading
        - croak if required args namespace or dsn not specified

0.15  Thu Jan 26 02:05:30 2006
        - tons of DB2-related fixes
        - better schema support for postgres
        - bugfix to sqlite pk-finding
        - test updates/fixes
        - fixes for injecting base classes
        - mysql quoting detection more reliable
        - DBIx::Class dependency bumped to 0.04001

0.14  Sat Jan 21 00:05:11 2006
        - Bugfix to pgsql tests
        - Bugfix for pause/cpan module version parsing

0.13  Thu Jan 19 06:40:11 2006
        - Test suite infrastructure overhauled
        - New advanced relationship testing [chisel]
        - Several relationship fixes
        - some minor pod cleanup
        - misc bugfixes

0.12  Fri Jan 13 06:09:04 2006
        - Changed "comment" to "Comment" for mysql "SHOW TABLE STATUS"
        - Added relationship tests for pg, mysql, and db2

0.11  Mon Jan 02 17:25:14 2006
        - use UNIVERSAL::require for loading the implementation class [Jason Kohles]
        - docs fixes [Jason Kohles]
        - fixed http://rt.cpan.org/NoAuth/Bug.html?id=16100 (implicit disconnects) [blblack]
        - carp rather than croak if a table has no PKs [blblack]
        - Added test for DB2 [blblack]

0.10  Tue Dec 20 06:22:23 2005
        - Fixed foreign key relationships to columns other than primary keys [blblack]
        - Fixed mysql test script [blblack]
        - Some schema support added [castaway]
        - DB2 support added [castaway]

0.09  Mon Nov 07 18:00:00 2005
        - Updated mysql loader

0.08  Mon Nov 07 18:00:00 2005
        - Fixed Pg loader

0.07  Mon Nov 07 18:00:00 2005
        - Fixed mysql loader

0.06  Thu Nov 03 18:00:00 2005
        - Fixed the _croak bug

0.05  Wed Nov 02 18:00:00 2005
        - Fixed dump output

0.04  Wed Nov 02 18:00:00 2005
        - Fixed Pg loader

0.03  Wed Nov 02 18:00:00 2005
        - new version number :P

0.02  Mon Oct 24 18:00:00 2005
        - Added $DBIx::Class::Loader::SCHEMA
        - Fixed Pg loader
        - Bumped version dependency.

0.01  Mon Sep 19 00:00:00 2005
        - original version.


t/dbixcl_common_tests.pm
package dbixcl_common_tests;

use strict;

use Test::More;
use DBIx::Class::Loader;
use DBI;

sub new {
    my $class = shift;

    my $self;

    if( ref($_[0]) eq 'HASH') {
       my $args = shift;
       $self = { (%$args) };
    }
    else {
       $self = { @_ };
    }

    # Only MySQL uses this
    $self->{innodb} ||= '';

    $self->{verbose} = $ENV{TEST_VERBOSE} || 0;

    return bless $self => $class;
}

sub skip_tests {
    my ($self, $why) = @_;

    plan skip_all => $why;
}

sub run_tests {
    my $self = shift;

    plan tests => 26;

    $self->create();

    my $namespace = 'DBIXCL_Test_' . $self->{vendor};

    my $debug = ($self->{verbose} > 1) ? 1 : 0;

    my %loader_opts = (
        dsn           => $self->{dsn},
        user          => $self->{user},
        password      => $self->{password},
        namespace     => $namespace,
        constraint    => '^(?:\S+\.)?(?i:loader_test)[0-9]+$',
        relationships => 1,
        debug         => $debug,
    );

    $loader_opts{schema} = $self->{schema} if $self->{schema};
    $loader_opts{dropschema} = $self->{dropschema} if $self->{dropschema};

    my $loader = DBIx::Class::Loader->new(%loader_opts);

    my $class1 = $loader->find_class("loader_test1");
    my $class2 = $loader->find_class("loader_test2");

    is( $class1, "${namespace}::LoaderTest1" );
    is( $class2, "${namespace}::LoaderTest2" );

    my $obj    = $class1->find(1);
    is( $obj->id,  1 );
    is( $obj->dat, "foo" );
    is( $class2->count, 4 );

    my ($obj2) = $class2->find( dat => 'bbb' );
    is( $obj2->id, 2 );

    SKIP: {
        skip $self->{skip_rels}, 20 if $self->{skip_rels};

        my $class3 = $loader->find_class("loader_test3");
        my $class4 = $loader->find_class("loader_test4");
        my $class5 = $loader->find_class("loader_test5");
        my $class6 = $loader->find_class("loader_test6");
        my $class7 = $loader->find_class("loader_test7");
        my $class8 = $loader->find_class("loader_test8");
        my $class9 = $loader->find_class("loader_test9");

        is( $class3, "${namespace}::LoaderTest3" );
        is( $class4, "${namespace}::LoaderTest4" );
        is( $class5, "${namespace}::LoaderTest5" );
        is( $class6, "${namespace}::LoaderTest6" );
        is( $class7, "${namespace}::LoaderTest7" );
        is( $class8, "${namespace}::LoaderTest8" );
        is( $class9, "${namespace}::LoaderTest9" );

        # basic rel test
        my $obj4 = $class4->find(123);
        is( ref($obj4->fkid), $class3);

        # fk def in comments should not be parsed
        my $obj5 = $class5->find( id1 => 1, id2 => 1 );
        is( ref( $obj5->id2 ), '' );

        # mulit-col fk def (works halfway for some, not others...)
        my $obj6   = $class6->find(1);
        isa_ok( $obj6->loader_test2, $class2 );
        SKIP: {
            skip "Multi-column FKs are only half-working for this vendor", 1
                unless $self->{multi_fk_broken};
            is( ref( $obj6->id2 ), '' );
        }

        # fk that references a non-pk key (UNIQUE)
        my $obj8 = $class8->find(1);
        isa_ok( $obj8->loader_test7, $class7 );

        # from Chisel's tests...
        SKIP: {
            if($self->{vendor} =~ /sqlite/i) {
                skip 'SQLite cannot do the advanced tests', 8;
            }

            my $class10 = $loader->find_class('loader_test10');
            my $class11 = $loader->find_class('loader_test11');

            is( $class10, "${namespace}::LoaderTest10" ); 
            is( $class11, "${namespace}::LoaderTest11" );

            my $obj10 = $class10->create({ subject => 'xyzzy' });

            $obj10->update();
            ok( defined $obj10, '$obj10 is defined' );

            my $obj11 = $class11->create({ loader_test10 => $obj10->id() });
            $obj11->update();
            ok( defined $obj11, '$obj11 is defined' );

            eval {
                my $obj10_2 = $obj11->loader_test10;
                $obj10_2->loader_test11( $obj11->id11() );
                $obj10_2->update();
            };
            is($@, '', 'No errors after eval{}');

            SKIP: {
                skip 'Previous eval block failed', 3
                    unless ($@ eq '');
        
                my $results = $class10->search({ subject => 'xyzzy' });
                is( $results->count(), 1,
                    'One $class10 returned from search' );

                my $obj10_3 = $results->first();
                isa_ok( $obj10_3, $class10 );
                is( $obj10_3->loader_test11()->id(), $obj11->id(),
                    'found same $class11 object we expected' );
            }

            for ( $class10, $class11 ) {
                $_->storage->dbh->disconnect;
            }
        }

        for ( $class3, $class4, $class5, $class6, $class7,
              $class8, $class9 ) {
            $_->storage->dbh->disconnect;
        }
    }

    for ( $class1, $class2 ) {
        $_->storage->dbh->disconnect;
    }
}

sub dbconnect {
    my ($self, $complain) = @_;

    my $dbh = DBI->connect(
         $self->{dsn}, $self->{user},
         $self->{password},
         {
             RaiseError => $complain,
             PrintError => $complain,
             AutoCommit => 1,
         }
    );

    die "Failed to connect to database: $DBI::errstr" if !$dbh;

    return $dbh;
}

sub create {
    my $self = shift;

    my @statements = (
        qq{
            CREATE TABLE loader_test1 (
                id $self->{auto_inc_pk},
                dat VARCHAR(32)
            ) $self->{innodb};
        },

        q{ INSERT INTO loader_test1 (dat) VALUES('foo'); },
        q{ INSERT INTO loader_test1 (dat) VALUES('bar'); }, 
        q{ INSERT INTO loader_test1 (dat) VALUES('baz'); }, 

        qq{ 
            CREATE TABLE loader_test2 (
                id $self->{auto_inc_pk},
                dat VARCHAR(32)
            ) $self->{innodb};
        },

        q{ INSERT INTO loader_test2 (dat) VALUES('aaa'); }, 
        q{ INSERT INTO loader_test2 (dat) VALUES('bbb'); }, 
        q{ INSERT INTO loader_test2 (dat) VALUES('ccc'); }, 
        q{ INSERT INTO loader_test2 (dat) VALUES('ddd'); }, 
    );

    my @statements_reltests = (
        qq{
            CREATE TABLE loader_test3 (
                id INTEGER NOT NULL PRIMARY KEY,
                dat VARCHAR(32)
            ) $self->{innodb};
        },

        q{ INSERT INTO loader_test3 (id,dat) VALUES(1,'aaa'); }, 
        q{ INSERT INTO loader_test3 (id,dat) VALUES(2,'bbb'); }, 
        q{ INSERT INTO loader_test3 (id,dat) VALUES(3,'ccc'); }, 
        q{ INSERT INTO loader_test3 (id,dat) VALUES(4,'ddd'); }, 

        qq{
            CREATE TABLE loader_test4 (
                id INTEGER NOT NULL PRIMARY KEY,
                fkid INTEGER NOT NULL,
                dat VARCHAR(32),
                FOREIGN KEY (fkid) REFERENCES loader_test3 (id)
            ) $self->{innodb};
        },

        q{ INSERT INTO loader_test4 (id,fkid,dat) VALUES(123,1,'aaa'); },
        q{ INSERT INTO loader_test4 (id,fkid,dat) VALUES(124,2,'bbb'); }, 
        q{ INSERT INTO loader_test4 (id,fkid,dat) VALUES(125,3,'ccc'); },
        q{ INSERT INTO loader_test4 (id,fkid,dat) VALUES(126,4,'ddd'); },

        qq{
            CREATE TABLE loader_test5 (
                id1 INTEGER NOT NULL,
                id2 INTEGER NOT NULL, -- , id2 INTEGER REFERENCES loader_test1,
                dat VARCHAR(8),
                PRIMARY KEY (id1,id2)
            ) $self->{innodb};
        },

        q{ INSERT INTO loader_test5 (id1,id2,dat) VALUES (1,1,'aaa'); },

        qq{
            CREATE TABLE loader_test6 (
                id $self->{auto_inc_pk},
                id2 INTEGER,
                loader_test2 INTEGER,
                dat VARCHAR(8),
                FOREIGN KEY (loader_test2) REFERENCES loader_test2 (id),
                FOREIGN KEY (id, id2 ) REFERENCES loader_test5 (id1,id2)
            ) $self->{innodb};
        },

        (q{ INSERT INTO loader_test6 (id2,loader_test2,dat) } .
         q{ VALUES (1,1,'aaa'); }),

        qq{
            CREATE TABLE loader_test7 (
                id INTEGER NOT NULL PRIMARY KEY,
                id2 VARCHAR(8) NOT NULL UNIQUE,
                dat VARCHAR(8)
            ) $self->{innodb};
        },

        q{ INSERT INTO loader_test7 (id,id2,dat) VALUES (1,'aaa','bbb'); },

        qq{
            CREATE TABLE loader_test8 (
                id INTEGER NOT NULL PRIMARY KEY,
                loader_test7 VARCHAR(8) NOT NULL,
                dat VARCHAR(8),
                FOREIGN KEY (loader_test7) REFERENCES loader_test7 (id2)
            ) $self->{innodb};
        },

        (q{ INSERT INTO loader_test8 (id,loader_test7,dat) } .
         q{ VALUES (1,'aaa','bbb'); }),

        qq{
            CREATE TABLE loader_test9 (
                loader_test9 VARCHAR(8) NOT NULL
            ) $self->{innodb};
        },
    );

    my @statements_advanced = (
        qq{
            CREATE TABLE loader_test10 (
                id10 $self->{auto_inc_pk},
                subject VARCHAR(8),
                loader_test11 INTEGER
            ) $self->{innodb};
        },

        qq{
            CREATE TABLE loader_test11 (
                id11 $self->{auto_inc_pk},
                message VARCHAR(8) DEFAULT 'foo',
                loader_test10 INTEGER,
                FOREIGN KEY (loader_test10) REFERENCES loader_test10 (id10)
            ) $self->{innodb};
        },

        (q{ ALTER TABLE loader_test10 ADD CONSTRAINT } .
         q{ loader_test11_fk FOREIGN KEY (loader_test11) } .
         q{ REFERENCES loader_test11 (id11); }),
    );

    $self->drop_tables;

    $self->{created} = 1;

    my $dbh = $self->dbconnect(1);
    $dbh->do($_) for (@statements);
    unless($self->{skip_rels}) {
        # hack for now, since DB2 doesn't like inline comments, and we need
        # to test one for mysql, which works on everyone else...
        # this all needs to be refactored anyways.
        if($self->{vendor} =~ /DB2/i) {
            @statements_reltests = map { s/--.*\n//; $_ } @statements_reltests;
        }
        $dbh->do($_) for (@statements_reltests);
        unless($self->{vendor} =~ /sqlite/i) {
            $dbh->do($_) for (@statements_advanced);
        }
    }
    $dbh->disconnect();
}

sub drop_tables {
    my $self = shift;

    return unless $self->{created};

    my @tables = qw/
        loader_test1
        loader_test2
    /;

    my @tables_reltests = qw/
        loader_test4
        loader_test3
        loader_test6
        loader_test5
        loader_test8
        loader_test7
        loader_test9
    /;

    my @tables_advanced = qw/
        loader_test11
        loader_test10
    /;

    my $drop_fk_mysql =
        q{ALTER TABLE loader_test10 DROP FOREIGN KEY loader_test11_fk;};

    my $drop_fk =
        q{ALTER TABLE loader_test10 DROP CONSTRAINT loader_test11_fk;};

    my $dbh = $self->dbconnect(0);

    unless($self->{skip_rels}) {
        $dbh->do("DROP TABLE $_") for (@tables_reltests);
        unless($self->{vendor} =~ /sqlite/i) {
            if($self->{vendor} =~ /mysql/i) {
                $dbh->do($drop_fk_mysql);
            }
            else {
                $dbh->do($drop_fk);
            }
            $dbh->do("DROP TABLE $_") for (@tables_advanced);
        }
    }
    $dbh->do("DROP TABLE $_") for (@tables);
    $dbh->disconnect;
}

sub DESTROY { shift->drop_tables; }

1;


lib/DBIx/Class/Loader/Pg.pm
package DBIx::Class::Loader::Pg;

use strict;
use base 'DBIx::Class::Loader::Generic';
use DBI;
use Carp;

=head1 NAME

DBIx::Class::Loader::Pg - DBIx::Class::Loader Postgres Implementation.

=head1 SYNOPSIS

  use DBIx::Class::Loader;

  # $loader is a DBIx::Class::Loader::Pg
  my $loader = DBIx::Class::Loader->new(
    dsn       => "dbi:Pg:dbname=dbname",
    user      => "postgres",
    password  => "",
    namespace => "Data",
  );
  my $class = $loader->find_class('film'); # $class => Data::Film
  my $obj = $class->retrieve(1);

=head1 DESCRIPTION

See L<DBIx::Class::Loader>.

=cut

sub _db_classes {
    return qw/DBIx::Class::PK::Auto::Pg/;
}

sub _tables {
    my $self = shift;

    $self->{_schema} ||= 'public'; # XXX this should be in an override of ->new().
                                   # the issue is that ::Generic actually does all
                                   # of its work in ->new(), rather than later.

    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);

    # This is split out to avoid version parsing errors...
    my $is_dbd_pg_gte_131 = ( $DBD::Pg::VERSION >= 1.31 );
    my @tables = $is_dbd_pg_gte_131 ? 
        $dbh->tables( undef, $self->{_schema}, "", "table", { noprefix => 1, pg_noprefix => 1 } )
        : $dbh->tables;

    $dbh->disconnect;
    s/"//g for @tables;
    return @tables;
}

sub _table_info {
    my ( $self, $table ) = @_;
    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);

    my $sth = $dbh->column_info(undef, $self->{_schema}, $table, undef);
    my @cols = map { $_->[3] } @{ $sth->fetchall_arrayref };
    s/"//g for @cols;
    
    my @primary = $dbh->primary_key(undef, $self->{_schema}, $table);

    $dbh->disconnect;

    s/"//g for @primary;

    return ( \@cols, \@primary );
}

=head1 SEE ALSO

L<DBIx::Class::Loader>

=cut

1;


lib/DBIx/Class/Loader/DB2.pm
package DBIx::Class::Loader::DB2;

use strict;
use base 'DBIx::Class::Loader::Generic';
use DBI;
use Carp;

=head1 NAME

DBIx::Class::Loader::DB2 - DBIx::Class::Loader DB2 Implementation.

=head1 SYNOPSIS

  use DBIx::Class::Loader;

  # $loader is a DBIx::Class::Loader::DB2
  my $loader = DBIx::Class::Loader->new(
    dsn       => "dbi:DB2:dbname",
    user      => "myuser",
    password  => "",
    namespace => "Data",
    schema    => "MYSCHEMA",
    dropschema  => 0,
  );
  my $class = $loader->find_class('film'); # $class => Data::Film
  my $obj = $class->retrieve(1);

=head1 DESCRIPTION

See L<DBIx::Class::Loader>.

=head1 KNOW ISSUES

L<DBIx::Class::Loader::DB2> will not pass test suites (nor work quite
completely for applications) with L<DBIx::Class> versions less than
0.05.  This is because prior to that release, there was no
L<DBIx::Class::PK::Auto::DB2>.  As long as you don't use any
auto-incrementing primary keys, things should be sane though.

=cut

sub _db_classes {
    return qw/DBIx::Class::PK::Auto::DB2/;
}

sub _tables {
    my $self = shift;
    my %args = @_; 
    my $schema = uc ($args{schema} || '');
    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);

    # this is split out to avoid version parsing errors...
    my $is_dbd_db2_gte_114 = ( $DBD::DB2::VERSION >= 1.14 );
    my @tables = $is_dbd_db2_gte_114 ? 
    $dbh->tables( { TABLE_SCHEM => '%', TABLE_TYPE => 'TABLE,VIEW' } )
        : $dbh->tables;
    $dbh->disconnect;

    # People who use table or schema names that aren't identifiers deserve
    # what they get.  Still, FIXME?
    s/\"//g for @tables;
    @tables = grep {!/^SYSIBM\./ and !/^SYSCAT\./ and !/^SYSSTAT\./} @tables;
    @tables = grep {/^$schema\./} @tables if($schema);
    return @tables;
}

sub _table_info {
    my ( $self, $table ) = @_;
#    $|=1;
#    print "_table_info($table)\n";
    my ($schema, $tabname) = split /\./, $table, 2;
    # print "Schema: $schema, Table: $tabname\n";
    
    # FIXME: Horribly inefficient and just plain evil. (JMM)
    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);
    $dbh->{RaiseError} = 1;

    my $sth = $dbh->prepare(<<'SQL') or die;
SELECT c.COLNAME
FROM SYSCAT.COLUMNS as c
WHERE c.TABSCHEMA = ? and c.TABNAME = ?
SQL

    $sth->execute($schema, $tabname) or die;
    my @cols = map { lc $_ } map { @$_ } @{$sth->fetchall_arrayref};
    $sth->finish;

    $sth = $dbh->prepare(<<'SQL') or die;
SELECT kcu.COLNAME
FROM SYSCAT.TABCONST as tc
JOIN SYSCAT.KEYCOLUSE as kcu ON tc.constname = kcu.constname
WHERE tc.TABSCHEMA = ? and tc.TABNAME = ? and tc.TYPE = 'P'
SQL

    $sth->execute($schema, $tabname) or die;

    my @pri = map { lc $_ } map { @$_ } @{$sth->fetchall_arrayref};

    $sth->finish;
    $dbh->disconnect;
    
    return ( \@cols, \@pri );
}

# Find and setup relationships
sub _relationships {
    my $self = shift;

    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);
    $dbh->{RaiseError} = 1;
    my $sth = $dbh->prepare(<<'SQL') or die;
SELECT SR.COLCOUNT, SR.REFTBNAME, SR.PKCOLNAMES, SR.FKCOLNAMES
FROM SYSIBM.SYSRELS SR WHERE SR.TBNAME = ?
SQL

    foreach my $table ( $self->tables ) {
        if ($sth->execute(uc $table)) {
            while(my $res = $sth->fetchrow_arrayref()) {
                my ($colcount, $other, $other_column, $column) =
                    map { $_=lc; s/^\s+//; s/\s+$//; $_; } @$res;
                next if $colcount != 1; # XXX no multi-col FK support yet
                eval { $self->_belongs_to_many( $table, $column, $other,
                  $other_column ) };
                warn qq/\# belongs_to_many failed "$@"\n\n/
                  if $@ && $self->debug;
            }
        }
    }

    $sth->finish;
    $dbh->disconnect;
}


=head1 SEE ALSO

L<DBIx::Class::Loader>

=cut

1;


lib/DBIx/Class/Loader/mysql.pm
package DBIx::Class::Loader::mysql;

use strict;
use base 'DBIx::Class::Loader::Generic';
use DBI;
use Carp;

=head1 NAME

DBIx::Class::Loader::mysql - DBIx::Class::Loader mysql Implementation.

=head1 SYNOPSIS

  use DBIx::Class::Loader;

  # $loader is a DBIx::Class::Loader::mysql
  my $loader = DBIx::Class::Loader->new(
    dsn       => "dbi:mysql:dbname",
    user      => "root",
    password  => "",
    namespace => "Data",
  );
  my $class = $loader->find_class('film'); # $class => Data::Film
  my $obj = $class->retrieve(1);

=head1 DESCRIPTION

See L<DBIx::Class::Loader>.

=cut

sub _db_classes {
    return qw/DBIx::Class::PK::Auto::MySQL/;
}

# Very experimental and untested!
sub _relationships {
    my $self   = shift;
    my @tables = $self->tables;
    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);

    foreach my $table (@tables) {
        my $query = "SHOW CREATE TABLE ${table}";
        my $sth   = $dbh->prepare($query)
          or die("Cannot get table definition: $table");
        $sth->execute;
        my $table_def = $sth->fetchrow_arrayref->[1] || '';
        
        my (@cols) = ($table_def =~ /CONSTRAINT `.*` FOREIGN KEY \(`(.*)`\) REFERENCES `(.*)` \(`(.*)`\)/g);

        while (scalar @cols > 0) {
            my $column = shift @cols;
            my $remote_table = shift @cols;
            my $remote_column = shift @cols;
            
            eval { $self->_belongs_to_many( $table, $column, $remote_table, $remote_column) };
            warn qq/\# belongs_to_many failed "$@"\n\n/ if $@ && $self->debug;
        }
        
        $sth->finish;
    }
}

sub _tables {
    my $self = shift;
    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);
    my @tables;
    my $quoter = $dbh->get_info(29) || q{`};
    foreach my $table ( $dbh->tables ) {
        $table =~ s/$quoter//g if ($quoter);
        push @tables, $1
          if $table =~ /\A(\w+)\z/;
    }
    $dbh->disconnect;
    return @tables;
}

sub _table_info {
    my ( $self, $table ) = @_;
    my $dbh = DBI->connect( @{ $self->{_datasource} } ) or croak($DBI::errstr);

    # MySQL 4.x doesn't support quoted tables
    my $query = "DESCRIBE $table";
    my $sth = $dbh->prepare($query) or die("Cannot get table status: $table");
    $sth->execute;
    my ( @cols, @pri );
    while ( my $hash = $sth->fetchrow_hashref ) {
        my ($col) = $hash->{Field} =~ /(\w+)/;
        push @cols, $col;
        push @pri, $col if $hash->{Key} eq "PRI";
    }

    $dbh->disconnect;
    return ( \@cols, \@pri );
}

=head1 SEE ALSO

L<DBIx::Class::Loader>

=cut

1;


lib/DBIx/Class/Loader/Writing.pm
package DBIx::Class::Loader::Writing;

# Empty. POD only.

1;

=head1 NAME                                                                     
                                                                                
DBIx::Class::Loader::Writing - Loader subclass writing guide

=head1 SYNOPSIS

  package DBIx::Class::Loader::Foo;

  # THIS IS JUST A TEMPLATE TO GET YOU STARTED.

  use strict;
  use base 'DBIx::Class::Loader::Generic';
  use DBI;
  use Carp;

  sub _db_classes {
      return qw/DBIx::Class::PK::Auto::Foo/;
          # You may want to return more, or less, than this.
  }

  sub _tables {
      my $self = shift;
      my $dbh = DBI->connect( @{ $self->{_datasource} } )
           or croak($DBI::errstr);
      return $dbh->tables; # Your DBD may need something different
  }

  sub _table_info {
      my ( $self, $table ) = @_;
      ...
      return ( \@cols, \@primary );
   }

   sub _relationships {
      my $self = shift;
      ...
      $self->_belongs_to_many($table, $f_key, $f_table);
          # For each relationship you want to set up
      ...
    }

=cut


lib/DBIx/Class/Loader.pm
package DBIx::Class::Loader;

use strict;
use Carp;
use UNIVERSAL::require;

our $VERSION = '0.20';

=head1 NAME

DBIx::Class::Loader - Dynamic definition of DBIx::Class sub classes.

=head1 SYNOPSIS

  use DBIx::Class::Loader;

  my $loader = DBIx::Class::Loader->new(
    dsn                     => "dbi:mysql:dbname",
    user                    => "root",
    password                => "",
    namespace               => "Data",
    additional_classes      => [qw/DBIx::Class::Foo/],
    additional_base_classes => [qw/My::Stuff/],
    left_base_classes       => [qw/DBIx::Class::Bar/],
    constraint              => '^foo.*',
    relationships           => 1,
    options                 => { AutoCommit => 1 }, 
    inflect                 => { child => 'children' },
    debug                   => 1,
  );
  my $class = $loader->find_class('film'); # $class => Data::Film
  my $obj = $class->find(1);

use with mod_perl

in your startup.pl

  # load all tables
  use DBIx::Class::Loader;
  my $loader = DBIx::Class::Loader->new(
    dsn       => "dbi:mysql:dbname",
    user      => "root",
    password  => "",
    namespace => "Data",
  );

in your web application.

  use strict;

  # you can use Data::Film directly
  my $film = Data::Film->retrieve($id);

=head1 IMPORTANT NOTICE

This module is deprecated in favor of L<DBIx::Class::Schema::Loader> for
use with L<DBIx::Class> versions 0.05 and higher.  It continues to
function as well as it ever did, even for recent L<DBIx::Class> releases,
and will be maintained for some time to counter bugs, but it doesn't use
the now-preferred L<DBIx::Class::Schema> way of doing things, and tends
to promote bad L<DBIx::Class> usage habits.

=head1 DESCRIPTION

DBIx::Class::Loader automate the definition of DBIx::Class sub-classes by
scanning table schemas and setting up columns and primary keys.

Class names are defined by table names and the namespace option.  The only
required arguments are C<namespace> and C<dsn>.

 +---------+-----------+--------------+
 | table   | namespace | class        |
 +---------+-----------+--------------+
 | foo     | Data      | Data::Foo    |
 | foo_bar | MyDB      | MyDB::FooBar |
 +---------+-----------+--------------+

DBIx::Class::Loader supports MySQL, Postgres, SQLite and DB2.  See
L<DBIx::Class::Loader::Generic> for more, and L<DBIx::Class::Loader::Writing>
for notes on writing your own db-specific subclass for an unsupported db.

=cut

=head1 METHODS

=head2 new

Example in Synopsis above demonstrates the available arguments.  For
detailed information on the arguments, see the
L<DBIx::Class::Loader::Generic> documentation.

=cut

sub new {
    my ( $class, %args ) = @_;

    foreach (qw/dsn namespace/) {
        croak qq/Argument '$_' is required/ unless defined($args{$_});
    }

    my $dsn = $args{dsn};

    my ($driver) = $dsn =~ m/^dbi:(\w*?)(?:\((.*?)\))?:/i;
    $driver = 'SQLite' if $driver eq 'SQLite2';
    my $impl = "DBIx::Class::Loader::" . $driver;
    $impl->require or
      croak qq/Couldn't require loader class "$impl"/
          . qq/, "$UNIVERSAL::require::ERROR"/;

    return $impl->new(%args);
}

=head1 SUPPORT

Bug reports to Brandon L Black C<blblack@gmail.com>,
or the mailing list dbix-class@lists.rawmode.org,
or visit #dbix-class on irc.perl.org.

=head1 AUTHOR

Sebastian Riedel, C<sri@oook.de>

Based upon the work of IKEBE Tomohiro

=head1 THANK YOU

Adam Anderson, Andy Grundman, Autrijus Tang, Dan Kubb, David Naughton,
Randal Schwartz, Simon Flack and all the others who've helped.

=head1 LICENSE

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=head1 SEE ALSO

L<DBIx::Class>

=cut

1;
